web/src/message_helper.ts

```diff
@@ -15,107 +15,174 @@ import * as stream_topic_history from "./stream_topic_history.ts";
-            util.set_match_data(cached_msg, raw_message);
+        if (
+            opts.type === "server_message" &&
+            util.get_match_topic(opts.raw_message) !== undefined
+        ) {
+            util.set_match_data(cached_msg, opts.raw_message);
         }
         return cached_msg;
     }
 
-    const message_with_booleans =
-        message_store.convert_raw_message_to_message_with_booleans(raw_message);
-    people.extract_people_from_message(message_with_booleans);
+    const message_with_booleans = message_store.convert_raw_message_to_message_with_booleans(opts);
+    people.extract_people_from_message(message_with_booleans.message);
 
-    const sent_by_me = people.is_my_user_id(message_with_booleans.sender_id);
-    people.maybe_incr_recipient_count({...message_with_booleans, sent_by_me});
+    const sent_by_me = people.is_my_user_id(message_with_booleans.message.sender_id);
+    people.maybe_incr_recipient_count({...message_with_booleans.message, sent_by_me});
 
     let status_emoji_info;
-    const sender = people.maybe_get_user_by_id(message_with_booleans.sender_id);
+    const sender = people.maybe_get_user_by_id(message_with_booleans.message.sender_id);
     if (sender) {
-        message_with_booleans.sender_full_name = sender.full_name;
-        message_with_booleans.sender_email = sender.email;
-        status_emoji_info = user_status.get_status_emoji(message_with_booleans.sender_id);
-    }
-
-    // TODO: Remove this once we've converted more message code to typescript
-    // and also have confirmed we don't see this error for some period of time.
-    if (!raw_message.reactions) {
-        blueslip.error("expected raw_message to have reactions");
-        raw_message.reactions = [];
+        message_with_booleans.message.sender_full_name = sender.full_name;
+        message_with_booleans.message.sender_email = sender.email;
+        status_emoji_info = user_status.get_status_emoji(message_with_booleans.message.sender_id);
     }
 
     // TODO: Rather than adding this field to the message object, it
     // might be cleaner to create an independent map from message_id
     // => clean_reactions data for the message, with care being taken
     // to make sure reify_message_id moves the data structure
     // properly.
-    const clean_reactions = reactions.generate_clean_reactions(raw_message);
+    const clean_reactions = reactions.generate_clean_reactions(opts.raw_message);
 
     let message: Message;
-    if (message_with_booleans.type === "stream") {
-        const topic = message_with_booleans.topic ?? message_with_booleans.subject;
+    if (message_with_booleans.message.type === "stream") {
+        let topic;
+        if (message_with_booleans.type === "local_message") {
+            topic = message_with_booleans.message.topic;
+        } else {
+            topic = message_with_booleans.message.topic ?? message_with_booleans.message.subject;
```

Comment:
I feel like we should be able to do just `message_with_booleans.message.subject` for this else clause -- since it's `subject` in the API still.